### ISSUE TITLE
"Policy collapse" issue

### ISSUE THREAD

**Issue by ruanrothmann** [Unity-Technologies/ml-agents#1336]
In some of my continuous control training setups I keep getting an odd issue and I'd like to hear if this is something anyone else has encountered and if there are any ideas someone might have for troubleshooting or solutions.

The agent makes good progress up until a certain point, at which the policy seems to get corrupted in some way. Here are some tensorboard graphs showing the issue: https://imgur.com/a/HG7Hbc4

Note the massive sudden dives in cumulative reward, accompanied by big spikes in policy and value loss. Has anyone here encountered this, and/or have any ideas for how I can go about figuring out the cause?  It's interesting to note that training does seem to recover in some way - after the 'collapse' it reaches the previous level faster than it did initially. However, at some stage it seems that the collapse is final and deadly.

My first instinct is that observed values, or reward, might be some huge (or NaN) value in some steps that causes gradient descent to break, but I carefully went over the code and I'm also manually clamping each observed value to some reasonable scale, but this does not seem to solve the issue.

An aside question - the convention in UnityML is calculating reward as part of the action. To me, it makes more sense to collect reward during observation, any reason for working it that way around?

**Comment by Serious-Joker** [comment]
https://youtu.be/8EcdaCk9KaQ?t=1731
This might be helpful to diagnose your issue. Try lowering your learning rate/step size. Your algorithm might be overstepping.

**Comment by StrSpi** [comment]
I also have this problem. I make an environment to help the agent find a goal. The agent can reach the target at the last episode, but can only rotate in place at the next episode. Mean reward went down dramatically, and could not be recovered.

**Comment by ruanrothmann** [comment]
I (mostly) resolved my issue by decreasing epsilon, epochs and learning rate.  Something else to check is that learning environment is not degrading (e.g. a ragdoll explodes and does not get reprojected).  Larger batch/buffer sizes will probably also help address the issue.
